### DISCUSSION TITLE
How to determine which functions are the GTK equivalent of WinForms?

### DISCUSSION THREAD

**Discussion by PlatinumLucario** [GtkSharp/GtkSharp#331] [Q&A] [answered]
As I've been in the process of migrating specific UI functions to GTKSharp, I've been really curious as to which GTKSharp functions are the GTK equivalent of WinForms?

For example, I have this piece of code for Sanford.Multimedia.Midi, which I'm in the process of creating a GTK version of the piano key UI for:

```
protected override void OnMouseDown(MouseEventArgs e)
            {
                PressPianoKey();

                if(!owner.Focused)
                {
                    owner.Focus();
                }

                base.OnMouseDown(e);
            }
```

I've been struggling with creating a GTK equivalent of that code, so far I've only managed to get as far as this:

```
protected override void OnDragBegin(DragContext context)
            {
                PressPianoKey();


                if (OwnerChange)  //Error CS0119, due to it being a type
                {
                    owner.Focus(); //Error CS1061, due to no definition available for 'Focus'
                }

                base.OnDragBegin(context);
            }
```

Is there any way I can find out which GTK code works in the same way as the Windows Forms code? Thank you so much!

**Comment by lytico** [comment, marked as answer]
I don't know any framework that supports gtk & winforms with c#.
(Are you familiar with  WPF? then there would be [xwt](https://github.com/mono/xwt), and you could dive into the different implementations to compare.)
Better:
see @harry-cpp 's answer: see Eto.Forms to compare!

**Reply by PlatinumLucario** [reply]
I'm quite new to WPF, and xwt sounds like an awesome suggestion for comparing the differences.

Though my intentions is for me to replace the WinForms code entirely with GTKSharp code, which is why I'm asking if anyone knows the GTK equivalent.

**Reply by harry-cpp** [reply]
@lytico [Eto.Forms](https://github.com/picoe/Eto) is a lot more powerful than xwt, has WinForms like API, and supports WinForms, WPF, Gtk and Mac backends out of the box. I highly recommend it.

@PlatinumLucario You will have to read Gtk C documentation to figure out the equivalent stuff, for `OnMouseDown` its `ButtonPressed`.
